### PR TITLE
EOS-16655: Mini-Provisioner Interface Implementation (setup.yaml)

### PR DESCRIPTION
**Pull request by ypise**
Signed-off-by: Yashodhan Pise <yashodhan.pise@seagate.com>

### CODE REVIEW

File: srv/_modules/setup_conf.py [hunk 1]
```diff
@@ -17,37 +17,60 @@
...
+
+    replacement_url = __pillar__['provisioner']['common_config']['url']
+    ret_val = ''
+    with open(conf_file, 'r') as fd:
+        try:
+            yml_dict = yaml.safe_load(fd)
+
+            # This split is hard-coded as this is the input format expected
+            # during call from the sls file.
+            root_node = conf_key.split(':')[0]
```
Comment (nitin-seagate) [inline srv/_modules/setup_conf.py:54]: Delimiter "**:**" is driven by file type (may be we can accept this as argument - with default being "**:**")

Rename **root_node** to **component_setup**

File: srv/_modules/setup_conf.py [hunk 2]
```diff
@@ -17,37 +17,60 @@
...
+            config_stage = conf_key.split(':')[1]
```
Comment (nitin-seagate) [inline srv/_modules/setup_conf.py:55]: Rename **config_stage** to **component_interface**

File: srv/_modules/setup_conf.py [hunk 3]
```diff
@@ -17,37 +17,60 @@
...
+            cmd_path = yml_dict[root_node][config_stage]['cmd']
+            logger.debug(f"The command path from yaml file: {cmd_path}")
+
+            # Proceed to process args, only if command has been specified
+            if cmd_path and os.path.exists(cmd_path):
```
Comment (nitin-seagate) [inline srv/_modules/setup_conf.py]: We will not be checking on cmd_path buc checking if the command is supported (may be using helper) just verify exit codes for the same.

File: srv/_modules/setup_conf.py [hunk 4]
```diff
@@ -17,37 +17,60 @@
...
+                args = yml_dict[root_node][config_stage]['args']
```
Comment (nitin-seagate) [inline srv/_modules/setup_conf.py:67]: Rename **yml_dict** to **config_info**
Rename **args** to **ci_setup_args**

File: srv/_modules/setup_conf.py [hunk 5]
```diff
@@ -17,37 +17,60 @@
...
+    with open(conf_file, 'r') as fd:
+        try:
+            yml_dict = yaml.safe_load(fd)
+
+            # This split is hard-coded as this is the input format expected
+            # during call from the sls file.
+            root_node = conf_key.split(':')[0]
+            config_stage = conf_key.split(':')[1]
+            cmd_path = yml_dict[root_node][config_stage]['cmd']
+            logger.debug(f"The command path from yaml file: {cmd_path}")
```
Comment (nitin-seagate) [inline srv/_modules/setup_conf.py]: Change debug message to something like this:
**_Component Setup Cmd: csm_setup init_**

File: srv/_modules/setup_conf.py [hunk 6]
```diff
@@ -17,37 +17,60 @@
...
+            # Proceed to process args, only if command has been specified
+            if cmd_path and os.path.exists(cmd_path):
+                args = yml_dict[root_node][config_stage]['args']
+
+                # If args is a string, do nothing.
+                # If args is a list, join the elements into a string
+                if isinstance(args, list):
+                    args = ' '.join(args)
+                    args.replace("$URL", replacement_url)
+                    logger.debug(f"Arguments for command {cmd_path}: {args}")
```
Comment (nitin-seagate) [inline srv/_modules/setup_conf.py]: Change debug message to something like this:
***Component Setup Args: csm_setup init --config <provisioner-url>***

File: srv/_modules/setup_conf.py [hunk 7]
```diff
@@ -17,37 +17,60 @@
...
+
+                ret_val = cmd_path + " " + str(args)
+                logger.debug(f"{ret_val}")
```
Comment (nitin-seagate) [inline srv/_modules/setup_conf.py:77]: Display final CI setup command something like this:
**_Component Setup: csm_setup init --config json:///opt/seagate/cortx_configs/provisioner_cluster.json_**

File: srv/_modules/setup_conf.py [hunk 8]
```diff
@@ -17,37 +17,60 @@
...
+
+        except yaml.YAMLError as ymlerr:
+            # Oops, yaml file was not well formed
+            logger.debug(f"Error parsing yaml file {conf_file}:\n {ymlerr}")
```
Comment (nitin-seagate) [inline srv/_modules/setup_conf.py:81]: Don;t mention keyword YAML file anywhere in display/info messages (Let that be driven by file-name itself)
We can mention something like this:
**Error parsing component setup config - 'setup.yaml' : 'yaml_error'**

File: pillar/components/provisioner.sls [hunk 9]
```diff
@@ -0,0 +1,3 @@
+provisioner:
+  common_config:
+    url: json:///opt/seagate/cortx_configs/provisioner_cluster.json
```
Comment (nitin-seagate) [inline pillar/components/provisioner.sls]: Rename **url** to **confstore_url**

**Review by nitin-seagate — CHANGES_REQUESTED** [review]

File: pillar/components/provisioner.sls (hunk 9 above)
[the remark this answers is not in the document]
Comment (nitin-seagate) [reply pillar/components/provisioner.sls]: Rename **url** to **confstore_url**

**Review by nitin-seagate — CHANGES_REQUESTED** [review]

File: srv/_modules/files/samples/setup.yaml [hunk 10]
```diff
@@ -17,27 +17,24 @@
 
 test:
     post_install:
```
[the remark this answers is not in the document]
Comment (ypise) [reply srv/_modules/files/samples/setup.yaml:19]: This is a test file, to check functionality of Python module and doesn't test the sanity of setup.yaml.

File: srv/_modules/setup_conf.py (hunk 4 above)
Comment (ypise) [reply srv/_modules/setup_conf.py:73]: @nitin-seagate 
Why ci?
This is deployment stage, which might not be part of CI. CI here is misleading.

File: pillar/components/provisioner.sls (hunk 9 above)
[the remark this answers is not in the document]
Comment (ypise) [reply pillar/components/provisioner.sls]: Addressed.

**Comment by ypise** [comment]
@nitin-seagate 
I've addressed the review comments. Kindly resolve if you accept them.

**Comment by ypise** [comment]
> ci_setup_cmd

@nitin-seagate,
Why do we want to call it ci_setup_cmd?
This is not for CI but for configuration. CI is just going to enable automation during deployment phase. However, the command is going to be reused in multiple environments.
In my opinion this is not CI speific.

**Review by nitin-seagate — APPROVED** [review]
Agree about the point (Not CI specific)...
Overall looks great to me... Thanks

File: srv/_modules/setup_conf.py [hunk 11]
```diff
@@ -17,37 +17,85 @@
...
+
+    confstore_url = __pillar__['provisioner']['common_config']['confstore_url']
+    ret_val = ''
+    with open(conf_file, 'r') as fd:
+        try:
+            config_info = yaml.safe_load(fd)
+
+            # This split is hard-coded as this is the input format expected
+            # during call from the sls file.
+            component_setup = config_info[conf_key.split(':')[0]]
```
Comment (nitin-seagate) [inline srv/_modules/setup_conf.py:54]: Should we explore taking this delimeter as input argument (default  :)

**Review by nitin-seagate — APPROVED** [review]

**Pull request merged**